Fixed nodes2

iosonofabio opened the pull request:
Next iteration of PR #8 on support for fixed nodes. This PR deprecates #8.

This time, as discussed on the phone, I've used Optimiser.optimize_partition and similar Python functions as entry points for an additional optional argument `fixed_nodes`, which is `None` or a list of `bool` of size equal to the number of nodes in the graph (if the list has int or other `True`-able Python objects, that's also ok).

Then I adapted the CPython API files, the `Optimiser.h` header file to include a few overloaded calls for the same functions as above relying on an array of `false` in case no nodes are fixed.

Finally I changed the `Optimiser.cpp` file with the actual implementation:
1. I do not add fixed nodes to the intial queue in `move_nodes`
2. I do not add fixed nodes in the subsequent, fast local moving queue in `move_nodes`
3. I exclude fixed nodes from the iteration in `merge_nodes`
4. I propagate fixed nodes down to the collapsed graphs (if you contain at least a fixed node, you are fixed yourself) in `optimise_partition`, starting from the full list of fixed nodes for the initial graph.

Moreover, I checked that all tests pass locally and added one simple test for two nodes that collapse without fixing but do not with fixing. Bottomline everything looks good from this side.

A few more points as we discussed:
- `move_nodes_constrained` and `merge_nodes_constrained` are untouched.
- nothing is currently done about community renumbering, not even at the very end. I would like you to double check the code and then we can implement that final touch.

Let me know what you think.

vtraag commented:
Thanks! I'll check it out somewhere next week.

vtraag commented on tests/test_VertexPartition.py:
```diff
@@ -175,8 +175,10 @@ def test_Bipartite(self):
+    self.optimiser.optimise_partition_multiplex(
+            [partition, partition_0, partition_1],
+            layer_weights=[1, -1, -1],
+            fixed_nodes=None)
```
It shouldn't be necessary to add the `fixed_nodes` argument, right?

iosonofabio replied:
This is an inconsistency in the preexisting interface, since optimise_partition_multiplex does not deal with keyword arguments. I'll fix that to be consistent with the rest of the code, but did not want to start fixing bugs around without a quick discussion,

iosonofabio commented on include/Optimiser.h:
```diff
@@ -24,25 +24,27 @@ class Optimiser
-    double move_nodes(vector<MutableVertexPartition*> partitions, vector<double> layer_weights);
-    double move_nodes(vector<MutableVertexPartition*> partitions, vector<double> layer_weights, int consider_comms, int consider_empty_community);
+    double move_nodes(MutableVertexPartition* partition, vector<bool> fixed_nodes, int consider_comms);
+    double move_nodes(vector<MutableVertexPartition*> partitions, vector<double> layer_weights, vector<bool> fixed_nodes);
```
moved everything to `const&`

iosonofabio commented on src/Optimiser.cpp:
```diff
@@ -334,11 +345,17 @@ double Optimiser::move_nodes(MutableVertexPartition* partition)
 
 double Optimiser::move_nodes(MutableVertexPartition* partition, int consider_comms)
+{
+  vector<bool> fixed_nodes(partition->get_graph()->vcount());
```
Well the default is to fix nothing right? That's vanilla Leiden.

vtraag replied:
Ah, yes of course, sorry, was thinking the other way around.

iosonofabio commented on src/Optimiser.cpp:
```diff
@@ -63,10 +63,10 @@ double Optimiser::optimise_partition(MutableVertexPartition* partition)
 {
   #ifdef DEBUG
-    cerr << "void Optimiser::optimise_partition(vector<MutableVertexPartition*> partitions, vector<double> layer_weights)" << endl;
+    cerr << "void Optimiser::optimise_partition(vector<MutableVertexPartition*> partitions, vector<double> layer_weights, vector<size_t> fixed_nodes)" << endl;
```
fixed - good catch

iosonofabio commented on include/Optimiser.h:
```diff
@@ -24,25 +24,27 @@ class Optimiser
   public:
     Optimiser();
-    double optimise_partition(MutableVertexPartition* partition);
+    double optimise_partition(MutableVertexPartition* partition, const vector<bool> fixed_nodes);
```
done both, thanks

iosonofabio commented on src/Optimiser.cpp:
```diff
@@ -423,20 +446,26 @@ double Optimiser::move_nodes(vector<MutableVertexPartition*> partitions, vector<
   size_t nb_moves = 0;
 
+  // Fixed nodes are also stable nodes
+  vector<int> is_node_stable(n, false);
```
done. Also changed the type to `vector<bool>` in the constrained function for consistency, but of course there are no fixed nodes there.

vtraag replied:
Yes, good to also change it there indeed.

iosonofabio commented on src/Optimiser.cpp:
```diff
@@ -423,20 +446,26 @@ double Optimiser::move_nodes(vector<MutableVertexPartition*> partitions, vector<
-  {
-    vertex_order.push(*it_node);
-  }
+    if (!fixed_nodes[*it_node])
```
fixed via adding only fixed nodes to the queue

iosonofabio commented on src/Optimiser.cpp:
```diff
@@ -423,20 +446,26 @@ double Optimiser::move_nodes(vector<MutableVertexPartition*> partitions, vector<
   // But if we use a random order, we shuffle this order.
+  // Also, we skip fixed nodes from the queue for efficiency reasons
   vector<size_t> nodes = range(n);
   shuffle(nodes, &rng);
```
done

iosonofabio commented:
Hi Vincent,

I fixed all but the renumbering communities. I already changed all the calls to the deprecated function into `set_membership`. As of implementing the renumbering, I think a `map` would be best as only a few nodes could be fixed. Otherwise I could make a `vector<int>` that sets `-1` as the community for all the nonfixed nodes. Do you have a strong opinion on this?

iosonofabio commented:
> This seems to be correct, but I believe it should also be corrected here
> https://github.com/iosonofabio/leidenalg/blob/ad7afdb081566f76c07bf8915509f907e819e29a/src/MutableVertexPartition.cpp#L248

I think that signature is not setting the memberships, it is just the multilayer equivalent of `void renumber_communities()` which is not being deprecated. So I don't think we should touch it

vtraag commented:
It's actually calling two methods. It is calling `vector<size_t> MutableVertexPartition::renumber_communities(vector<MutableVertexPartition*> partitions)` and passing the result to `MutableVertexPartition::renumber_communities(vector<size_t> const& membership)`. It should become

     this->set_membership(MutableVertexPartition::renumber_communities(partitions));

iosonofabio commented:
you are right, I just pushed the change

vtraag commented:
By the way, you are right, it would make more sense to have the call signature `renumber_communities(map<size_t, size_t> const& original_fixed_membership)`.

iosonofabio commented:
Yes, just puished that, working on the implementation now

vtraag commented:
Great, thanks!

iosonofabio commented:
I added a call to renumber the communities according to the fixed nodes at the end of `optimise_partition`. I'm not sure whether you want to add it at the end of `move_nodes` or `merge_nodes`: the current implementation might need ordered `int`s to work...

vtraag commented:
Thanks, I'll check the renumbering implementation!

We don't want to renumber the communities to the fixed nodes after every iteration of `move_nodes` or `merged_nodes` when called from `optimise_partition`. It doesn't make any difference if it is done only once at the end, or during each iteration of `move_nodes`. 

However, the `move_nodes` function is also directly exposed to Python, in which case they should be renumbered. I would propose to add a `bool renumber_fixed_nodes` argument to `move_nodes` and `merge_nodes`. If `renumber_fixed_nodes == true` we renumber the communities, and hence this is how it should be called from the public Python interface. Internally from `optimise_partition` we can then call the function with `renumber_fixed_nodes == false`. The other call signature could then default to `renumber_fixed_nodes==true`.

iosonofabio commented:
Yep, that was my idea as well

vtraag commented on src/MutableVertexPartition.cpp:
```diff
@@ -306,6 +306,48 @@ vector<size_t> MutableVertexPartition::renumber_communities(vector<MutableVertex
+  size_t nb_comms = n_communities();
+
+  // Fill the community map with the original communities
+  vector<size_t> comm_map(nb_comms);
```
Perhaps a bit confusing to call this a `map`, given that the variable `membership` is actually a `map`. Perhaps simply label this something like `new_comm_id` similar to the other `renumber_communities` function.

iosonofabio replied:
done

vtraag commented on src/MutableVertexPartition.cpp:
```diff
@@ -306,6 +306,48 @@ vector<size_t> MutableVertexPartition::renumber_communities(vector<MutableVertex
  Renumber the communities using the original fixed membership vector. Notice
  that this doesn't ensure any property of the community numbers.
 *****************************************************************************/
+void MutableVertexPartition::renumber_communities(map<size_t, size_t> const& membership)
```
I would follow the same pattern as for `vector<size_t> MutableVertexPartition::renumber_communities(vector<MutableVertexPartition*> partitions)`. Here we don't need the multiple partitions, so we can simpy use `vector<size_t> MutableVertexPartition::renumber_communities(map<size_t, size_t> const& membership)`, which should then return the new `membership` vector. In the line 348, we can then call `set_membership` with this new membership.

vtraag commented on src/Optimiser.cpp:
```diff
@@ -99,6 +99,16 @@ double Optimiser::optimise_partition(vector<MutableVertexPartition*> partitions,
       throw Exception("Number of nodes are not equal for all graphs.");
 
+  // Get the map of original communities for fixed nodes
+  vector< map<size_t, size_t> > original_fixed_memberships(nb_layers);
```
I don't think we need this for all layers. Effectively, all layers have identical membership information, so we can stick to only a single `map<size_t, size_t>`.

iosonofabio replied:
done

vtraag commented on src/Optimiser.cpp:
```diff
@@ -325,16 +335,21 @@ double Optimiser::optimise_partition(vector<MutableVertexPartition*> partitions,
+  // where r is the number of communities. The exception is fixed
+  // nodes which should keep the numbers of the original communities
   q = 0.0;
   vector<size_t> membership = MutableVertexPartition::renumber_communities(partitions);
```
Here we should then renumber using `partitions[0]->renumber_communities(original_fixed_memberships)`. We could leave the original statement as well, so that the communities will in principle first be ordered, and then be renumbered to follow the original fixed nodes. In other words, we obtain

    vector<size_t> membership = MutableVertexPartition::renumber_communities(partitions);
    partitions[0]->set_membership(membership);
    membership = partitions[0]->renumber_communities(original_fixed_membership);

The subsequent `for` loop can then start at `layer = 1` of course.

iosonofabio replied:
done

vtraag replied:
I just realised that the last line of my proposal would not set the membership of course, so the `for` loop should still start at `layer = 0`.

iosonofabio replied:
Correct, let me fix this

iosonofabio replied:
done

vtraag commented on src/Optimiser.cpp:
```diff
@@ -325,16 +335,21 @@ double Optimiser::optimise_partition(vector<MutableVertexPartition*> partitions,
     q += partitions[layer]->quality()*layer_weights[layer];
+
+    // Restore partition numbers for fixed_nodes
+    partitions[layer]->renumber_communities(original_fixed_memberships[layer]);
```
We don't need this then.

iosonofabio replied:
done

iosonofabio commented:
done

vtraag commented on src/MutableVertexPartition.cpp:
```diff
@@ -306,6 +306,48 @@ vector<size_t> MutableVertexPartition::renumber_communities(vector<MutableVertex
+  // Set the new communities
+  for (size_t i = 0; i < this->graph->vcount(); i++)
+  {
+    _membership[i] = comm_map[_membership[i]];
```
We should here then not set the internal `membership` but a different, separate, vector that we then return.

iosonofabio replied:
I extract it from the class instead of returning, I think it makes no difference...

iosonofabio replied:
anyway, I finally changed the signature as you asked above, that works too

iosonofabio commented:
Everything should be fixed as per your review, it builds locally and passes tests

vtraag commented:
Thanks for all the changes by the way. I just wanted to let you know that I won't have time to look at this before next Tuesday unfortunately.

iosonofabio commented:
No problem Vincent. I like this kind of work so I can push hard but I don't mind the wait either :-)

vtraag commented:
I was just testing the code a little bit, and I get a segfault.

```python
import leidenalg
import igraph as ig

G = ig.Graph.Full(3)

partition = leidenalg.CPMVertexPartition(G, resolution_parameter=0.01)
partition.membership[0] = 2
print(partition.membership)
opt = leidenalg.Optimiser()
fixed_nodes = [True, False, False]
opt.optimise_partition(partition, fixed_nodes=fixed_nodes)

print(partition.membership)
```
I haven't looked at the bug further, perhaps you can investigate?

iosonofabio commented:
Yep will look into it 😊

iosonofabio commented:
Got it, I had to add a condition for `aggregate_further` in the optimization loop that stops if all collapsed nodes end are fixed. In that case, nothing can move anyway.

We were entering the `move_nodes` function and trying to shuffle an empty queue, apparently the random number generator needs some food to chew on and an empty vector is not ok.

iosonofabio commented:
Moreover, you can't use:
```python
partition.membership[0] = 2
```
Instead, you have to use
```python
partition.set_membership([2, 1, 2])
```
or initial memberships via the constructor
```python
partition = leidenalg.CPMVertexPartition(G, resolution_parameter=0.01, initial_membership=[2, 1, 2])
```
That's a limitation of the current Python/C API interface, `partition.membership` seems to be a read-only, return-by-value kind of attribute, which generates all kinds of subtle bugs like this.

In a way, aboiding with these types of issues is why I'm not keen on venturing into C-API-ing myself ;-)

vtraag commented:
Thanks! I'll take a look tomorrow. Indeed the membership vector is read only. The problem is that it has to be synchronised with the rest of the administration, which is why you cannot set it manually. I could/should use `move_node(0, 2)` instead. But the fact that I myself already quickly forgot to take note is indeed a sign that that should be improved. I'll see what I can do about that.

vtraag commented:
Hmm, perhaps I simply ought to change it to a tuple instead of a list.

iosonofabio commented:
Hi Vincent,

I would like to get this closed, I think the PR is ready for merging?

Thanks
Fabio

vtraag commented:
Yes, sorry, won't get around to it due to Ascension. I'll merge and close it next week.

vtraag commented:
Merged with `develop` instead of `master`, and will be included in the next release. New features are added to `develop`, `master` contains the current released version and bugfixes for that version.

Thanks for the PR, hope it will be of much use (to you and others)!

vtraag commented:
Hi Fabio, I'm planning to release this soon, including built wheels for all platforms. Should people cite https://doi.org/10.1101/820928 for this contribution? I will then add it to the docs.

iosonofabio commented:
Vincent! I was gonna write you today as I got another request for this! That would be awesome, thank you!

Shall I try and add the same thing to igraph?

Cheers,
Fabio

On Sat, Apr 25, 2020, at 03:32, Vincent Traag wrote:
> 
> 
> Hi Fabio, I'm planning to release this soon, including built wheels for 
> all platforms. Should people cite https://doi.org/10.1101/820928 for 
> this contribution? I will then add it to the docs.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/vtraag/leidenalg/pull/9#issuecomment-619149224>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAJFEAEQFHFPOLTHYPBYXBLROHEL3ANCNFSM4HKPECRQ>.
>

vtraag commented:
OK, will do, hope to deploy to PyPi still this evening.

Yes, if you want to add the same thing to the Leiden implementation in `igraph`, sure, go right ahead! The implementation might be a bit more complex though.

vtraag commented:
OK, new release is made: https://pypi.org/project/leidenalg/, all binary wheels are building and will be pushed to PyPI in the coming hour.

iosonofabio commented:
Thank you!!

On Sat, Apr 25, 2020, at 07:54, Vincent Traag wrote:
> 
> 
> OK, new release is made: https://pypi.org/project/leidenalg/, all 
> binary wheels are building and will be pushed to PyPI in the coming 
> hour.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/vtraag/leidenalg/pull/9#issuecomment-619254863>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAJFEAFMLQSWAQ7CJCKONQLROIDB7ANCNFSM4HKPECRQ>.
>
